dom.go

```diff
@@ -346,89 +481,201 @@ func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts
-	for i := len(h.children); i < len(prev.children); i++ {
-		prevChild := prev.children[i]
-		if prevChildList, ok := prevChild.(List); ok {
+func (h *HTML) removeChildren(prevChildren []ComponentOrHTML) {
+	for _, prevChild := range prevChildren {
+		if prevChildList, ok := prevChild.(keyedList); ok {
 			// Previous child was a list, so remove all DOM nodes in it.
-			prevChildList.remove(h.node)
+			prevChildList.remove(h)
 			continue
 		}
 		prevChildRender := extractHTML(prevChild)
 		if prevChildRender == nil {
 			continue
 		}
-		unmount(prevChild)
-		removeNode(prevChildRender.node)
+		h.removeChild(prevChildRender)
+	}
+}
+
+// firstChild returns the first child DOM node of this element.
+func (h *HTML) firstChild() jsObject {
+	if h == nil || h.node == nil {
+		return nil
+	}
+	return h.node.Get("firstChild")
+}
+
+// nextSibling returns the next sibling DOM node for this element.
+func (h *HTML) nextSibling() jsObject {
+	if h == nil || h.node == nil {
+		return nil
+	}
+	return h.node.Get("nextSibling")
+}
+
+// removeChild removes the provided child element from this element, and
+// triggers unmount handlers.
+func (h *HTML) removeChild(child *HTML) {
+	// If we're removing the current insert target, use the next
+	// sibling, if any.
+	if h.insertBeforeNode != nil && h.insertBeforeNode == child.node {
+		h.insertBeforeNode = h.insertBeforeNode.Get("nextSibling")
+	}
+	unmount(child)
+	if child.node == nil {
+		return
+	}
+	// Use the child's parent node here, in case our node is not a valid
+	// target by the time we're called.
+	child.node.Get("parentNode").Call("removeChild", child.node)
+}
+
+// appendChild appends a new child to this element.
+func (h *HTML) appendChild(child *HTML) {
+	h.node.Call("appendChild", child.node)
+}
+
+// insertBefore inserts the provided child before the provided DOM node. If the
+// DOM node is nil, the child will be appended instead.
+func (h *HTML) insertBefore(node jsObject, child *HTML) {
+	if node == nil {
+		h.appendChild(child)
+		return
 	}
+	h.node.Call("insertBefore", child.node, node)
 }
 
 // List represents a list of components or HTML.
 type List []ComponentOrHTML
 
-// reconcile reconciles the List against the DOM node in isolation. If
-// insertBefore is non-nil, the List elements will be inserted into the DOM
-// before that node.
-func (l List) reconcile(node jsObject, insertBefore *HTML, prev ComponentOrHTML) []Mounter {
-	nextHTML := &HTML{node: node, children: l}
-	switch c := prev.(type) {
-	case List:
-		return nextHTML.reconcileChildren(&HTML{node: node, children: c}, insertBefore)
-	default:
-		if prev == nil {
-			return nextHTML.reconcileChildren(&HTML{node: node}, insertBefore)
-		}
-		return nextHTML.reconcileChildren(&HTML{node: node, children: []ComponentOrHTML{prev}}, insertBefore)
-	}
+// WithKey wraps the List in a Keyer using the given key.
+func (l List) WithKey(key interface{}) ComponentOrHTML {
+	return keyedList{key: key, html: &HTML{children: l}}
```

Comment:
My only concern with this is that MarkupOrChild/ComponentOrHTML may end up with this `keyedList` in them, which makes the documentation inaccurate, amongst other things.

Reply:
Yes that is a good point actually...

Reply:
We could publish a `type KeyedList interface{}` value + update the docs of `MarkupOrChild`/`ComponentOrHTML`?

Then nothing else has to change.

Reply:
Actually, just make `keyedList` public with no exported fields / methods

Reply:
That was my inclination, too.